### PR TITLE
Enable embedding custom metadata in the commit journal

### PR DESCRIPTION
**Pull request by silveryfu**
This PR allows embedding a metadata field (`meta`) in the commit object at `zed lake load`. The `meta` field can include any ZSON/ZNG data such that external applications can embed any custom metadata in the commit journal and implement arbitrary data provenance and governance capabilities later.

For example,

Load data with `-meta` option:
`zed lake load -meta '{class: "sensitive"}' in.zson`

Filter commits by metadata:
`zed lake query "from pool@main:log | meta.class=='sensitive'"`

One can embed any ZSON/ZNG data, e.g.,
`zed lake load -meta '[{a:1}, {b:"hi there"}]' in.zson`

This PR does not do/check: allowing data to be arbitrarily large (that it can't fit in an HTTP header over the load endpoint with remote driver).

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -2,7 +2,6 @@ package api
 
 import (
 	"context"
-
```
Comment (nwt) [inline api/api.go:5]: We separate standard library imports from 3rd-party imports with a blank line, so keep this. (If configured to manage imports, `gopls` should include this.)

File: lake/commits/actions.go [hunk 2]
```diff
@@ -56,15 +57,20 @@ type Commit struct {
...
 
 func (c *Commit) CommitID() ksuid.KSUID {
 	return c.ID
 }
 
-func (c *Commit) String() string {
-	//XXX need to format Message field for single line
-	return fmt.Sprintf("COMMIT %s -> %s %s %s %s", c.ID, c.Parent, c.Date, c.Author, c.Message)
+func (c *Commit) AppMeta() zed.Value {
+	return c.Meta
```
Comment (nwt) [inline lake/commits/actions.go:73]: These method changes don't appear to be necessary for this PR, so remove them to minimize the diff.

File: api/api.go (hunk 1 above)
Comment (silveryfu) [reply api/api.go:5]: thanks Noah! done!

File: lake/commits/actions.go (hunk 2 above)
Comment (silveryfu) [reply lake/commits/actions.go]: Ah, done! I'll double-check diff in future. thanks!

File: lake/commits/actions.go [hunk 3]
```diff
@@ -67,6 +69,10 @@ func (c *Commit) String() string {
 	return fmt.Sprintf("COMMIT %s -> %s %s %s %s", c.ID, c.Parent, c.Date, c.Author, c.Message)
 }
 
+func (c *Commit) AppMeta() zed.Value {
+	return c.Meta
+}
+
```
Comment (nwt) [inline lake/commits/actions.go:75]: This is unused, so let's not add it.

File: lake/commits/object.go [hunk 4]
```diff
@@ -21,7 +22,7 @@ type Object struct {
 	Actions []Action    `zed:"actions"`
 }
 
-func NewObject(parent ksuid.KSUID, author, message string, retries int) *Object {
+func NewObject(parent ksuid.KSUID, author, message string, retries int, meta zed.Value) *Object {
```
Comment (nwt) [inline lake/commits/object.go:25]: Nit: Put this next to `author` and `message`.
```suggestion
func NewObject(parent ksuid.KSUID, author, message string, meta zed.Value, retries int) *Object {
```

**Review by nwt — COMMENTED** [review]
Forgot to say that this looks great overall! A few more tweaks and we'll merge it.

File: lake/commits/actions.go (hunk 3 above)
Comment (silveryfu) [reply lake/commits/actions.go]: nice catch! removed!

File: lake/commits/object.go (hunk 4 above)
Comment (silveryfu) [reply lake/commits/object.go]: makes sense and done! thank you! :)

**Comment by mccanne** [comment]
Nice work @silveryfu!

File: lake/branch.go [hunk 5]
```diff
@@ -89,6 +94,17 @@ func loadMessage(objects []data.Object) string {
 	return b.String()
 }
 
+func loadMeta(meta string) (zed.Value, error) {
+	if meta == "" {
+		return zed.Value{zed.TypeNull, nil}, nil
+	}
+	zv, err := zson.ParseValue(zed.NewContext(), meta)
+	if err != nil {
+		return zed.Missing, fmt.Errorf("cannot parse zson string %s: %v", zv, err)
```
Comment (mattnibs) [inline lake/branch.go:103]: You should probably create a public error called `ErrInvalidCommitMeta` that looks like:
```
var ErrInvalidCommitMeta = errors.New("cannot parse zson string")
```
Then the `handleBranchLoad` handler on line 316 of service/handlers.go should check for this error and return ErrInvalid so the error code will be 400 in the response.

```suggestion
		return zed.Missing, fmt.Errorf("%w %s: %v", ErrInvalidCommitMeta, zv, err)
```

Comment (silveryfu) [reply lake/branch.go:103]: Thanks Matt! This makes a lot sense! Done!

**Review by mattnibs — APPROVED** [review]
lgtm

File: lake/branch.go [hunk 6]
```diff
@@ -25,7 +26,10 @@ const (
 	maxMessageObjects = 10
 )
 
-var ErrCommitFailed = fmt.Errorf("exceeded max update attempts (%d) to branch tip: commit failed", maxCommitRetries)
+var (
+	ErrCommitFailed      = fmt.Errorf("exceeded max update attempts (%d) to branch tip: commit failed", maxCommitRetries)
+	ErrInvalidCommitMeta = errors.New("cannot parse zson string")
```
Comment (nwt) [inline lake/branch.go:31]:
```suggestion
	ErrInvalidCommitMeta = errors.New("cannot parse ZSON string")
```

**Review by nwt — COMMENTED** [review]
One last request before we merge this.

File: lake/branch.go (hunk 6 above)
Comment (silveryfu) [reply lake/branch.go]: Thanks Noah! Done!

**Pull request merged**